### ISSUE TITLE
[MNG-5868] Adding several times the same artifact via MavenProjectHelper (attachArtifact) keep adding to the List duplicate artifacts

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7565]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** opened **[MNG-5868](https://issues.apache.org/jira/browse/MNG-5868?redirect=false)** and commented

During the check of an issue MSHADE-195 i stumbled over several things...

If you take a look here and the log output excerpt:

```
INFO] Minimized 2341 -> 1293
[INFO] Minimized 3282 -> 2234
[INFO] Replacing original artifact with shaded artifact.
[INFO] Replacing /Users/kama/ws-git/apache/mshade/mshade-195/target/MSHADE-195-example-1-SNAPSHOT.jar with /Users/kama/ws-git/apache/mshade/mshade-195/target/MSHADE-195-example-1-SNAPSHOT-shaded.jar
[INFO] Replacing original source artifact with shaded source artifact.
[INFO] Replacing /Users/kama/ws-git/apache/mshade/mshade-195/target/MSHADE-195-example-1-SNAPSHOT-sources.jar with /Users/kama/ws-git/apache/mshade/mshade-195/target/MSHADE-195-example-1-SNAPSHOT-shaded-sources.jar
[INFO] Dependency-reduced POM written at: /Users/kama/ws-git/apache/mshade/mshade-195/dependency-reduced-pom.xml
[INFO]
[INFO] --- maven-install-plugin:2.5.2:install (default-install) @ MSHADE-195-example ---
[INFO] Installing /Users/kama/ws-git/apache/mshade/mshade-195/target/MSHADE-195-example-1-SNAPSHOT.jar to /Users/kama/.m2/repository/com/example/MSHADE-195-example/1-SNAPSHOT/MSHADE-195-example-1-SNAPSHOT.jar
[INFO] Installing /Users/kama/ws-git/apache/mshade/mshade-195/dependency-reduced-pom.xml to /Users/kama/.m2/repository/com/example/MSHADE-195-example/1-SNAPSHOT/MSHADE-195-example-1-SNAPSHOT.pom
[INFO] Installing /Users/kama/ws-git/apache/mshade/mshade-195/target/MSHADE-195-example-1-SNAPSHOT-sources.jar to /Users/kama/.m2/repository/com/example/MSHADE-195-example/1-SNAPSHOT/MSHADE-195-example-1-SNAPSHOT-sources.jar
[INFO] Installing /Users/kama/ws-git/apache/mshade/mshade-195/target/MSHADE-195-example-1-SNAPSHOT-sources.jar to /Users/kama/.m2/repository/com/example/MSHADE-195-example/1-SNAPSHOT/MSHADE-195-example-1-SNAPSHOT-sources.jar
[INFO] 
```

Install plugin tries to install two identical artifacts which will work for maven-install-plugin but would fail a deploy to repository manager (for releases) etc.

So after diving into the problem i found the following code in maven-core (MavenProject.java):

```java
    /**
     * Add or replace an artifact. This method is now deprecated. Use the @{MavenProjectHelper} to attach artifacts to a
     * project. In spite of the 'throws' declaration on this API, this method has never thrown an exception since Maven
     * 3.0.x. Historically, it logged and ignored a second addition of the same g/a/v/c/t. Now it replaces the file for
     * the artifact, so that plugins (e.g. shade) can change the pathname of the file for a particular set of
     * coordinates.
     *
     * @param artifact the artifact to add or replace.
     * @throws DuplicateArtifactAttachmentException
     */
    public void addAttachedArtifact( Artifact artifact )
        throws DuplicateArtifactAttachmentException
    {
        getAttachedArtifacts().add( artifact );
    }

    public List<Artifact> getAttachedArtifacts()
    {
        if ( attachedArtifacts == null )
        {
            attachedArtifacts = new ArrayList<>();
        }
        return attachedArtifacts;
    }
```

So taking a look into MavenProjectHelper.java and the implementation (DefaultMavenProjectHelper.java).

```java
/**
 * Add an attached artifact or replace the file for an existing artifact.
 *
 * @see MavenProject#addAttachedArtifact(org.apache.maven.artifact.Artifact)
 * @param project project reference.
 * @param artifact artifact to add or replace.
 */
public void attachArtifact( MavenProject project, Artifact artifact )
{
    project.addAttachedArtifact( artifact );
}
```

which means that there is not check if an artifacts is already attached.

---

**Affects:** 3.2.3

**Issue Links:**
- [MDEPLOY-254](https://issues.apache.org/jira/browse/MDEPLOY-254) Maven Deploy Plugin deploy jar twice : Maven 3.3.3
 (_**"fixes"**_)
- [MINSTALL-159](https://issues.apache.org/jira/browse/MINSTALL-159) Prevent adding several times the same file for install
 (_**"fixes"**_)
- [MSOURCES-121](https://issues.apache.org/jira/browse/MSOURCES-121) Check for duplicated addition of the same file
 (_**"fixes"**_)
- [MNG-5387](https://issues.apache.org/jira/browse/MNG-5387) Add ability to replace an artifact in mid-build
 (_**"is duplicated by"**_)
- [MNG-5939](https://issues.apache.org/jira/browse/MNG-5939) Problem doing release when sources are generate as well
 (_**"is duplicated by"**_)
- [MSHARED-820](https://issues.apache.org/jira/browse/MSHARED-820) Check for duplicated attachment of the same artifact
 (_**"is duplicated by"**_)
- [MNG-7636](https://issues.apache.org/jira/browse/MNG-7636) Partially revert MNG-5868 to restore backward compatibility (see MNG-7316)
 (_**"causes"**_)
- [MNG-5387](https://issues.apache.org/jira/browse/MNG-5387) Add ability to replace an artifact in mid-build

- [MSHADE-195](https://issues.apache.org/jira/browse/MSHADE-195) createSourcesJar with source:jar-no-fork causes sources.jar to be deployed twice, causing the build to fail

- [MNG-5384](https://issues.apache.org/jira/browse/MNG-5384) Declarative artifacts

- [MNG-7316](https://issues.apache.org/jira/browse/MNG-7316) REGRESSION: MavenProject.getAttachedArtifacts() is read-only
 (_**"breaks"**_)

**Remote Links:**
- [Page
](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=67636510)
- [GitHub Pull Request #63
](https://github.com/apache/maven-integration-testing/pull/63)
- [GitHub Pull Request #356
](https://github.com/apache/maven/pull/356)
- [GitHub Pull Request #358
](https://github.com/apache/maven/pull/358)

**Backported to:** [4.0.0-alpha-2](https://github.com/apache/maven/milestone/113?closed=1), [3.8.2](https://github.com/apache/maven/milestone/103?closed=1)

5 votes, 19 watchers

**Comment by jira-importer** [comment]
**[Christian Schulte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=schulte77)** commented

Can someone explain what is the correct behaviour here? The Javadoc comment for 'MavenProject.addAttachedArtifact' and 'MavenProjectHelper.attachArtifact()' is not consistent to what the code does. For example:

MavenProject.addAttachedArtifact:

```
/**
 * Add or replace an artifact. This method is now deprecated. Use the @{MavenProjectHelper} to attach artifacts to a
 * project. In spite of the 'throws' declaration on this API, this method has never thrown an exception since Maven
 * 3.0.x. Historically, it logged and ignored a second addition of the same g/a/v/c/t. Now it replaces the file for
 * the artifact, so that plugins (e.g. shade) can change the pathname of the file for a particular set of
 * coordinates.
 *
 * @param artifact the artifact to add or replace.
 * @throws DuplicateArtifactAttachmentException
 */
```

DefaultMavenProjectHelper.attachArtifact(MavenProject,Artifact):

```
/**
 * Add an attached artifact or replace the file for an existing artifact.
 *
 * @see MavenProject#addAttachedArtifact(org.apache.maven.artifact.Artifact)
 * @param project project reference.
 * @param artifact artifact to add or replace.
 */
```

The code does not replace anything. It just adds the artifact to the list. How to solve this issue? Should the methods be updated to match the documentation, or is the documentation wrong?

**Comment by jira-importer** [comment]
**[Christian Schulte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=schulte77)** commented

`@Karl` Heinz Marbaise: Are you sure the list of affected versions is corrent? I think the issue got introduced by [this commit](https://git-wip-us.apache.org/repos/asf?p=maven.git;a=commit;h=6cf9320942c34bc68205425ab696b1712ace9ba4). So only Maven > 3.2.2 should be affected.

**Comment by jira-importer** [comment]
**[Jason Mihalick](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jrmihalick)** commented

As I mentioned on MNG-5939, I confirmed that the problem does not occur on Maven 3.2.1 (I didn't have access to 3.2.2).  It definitely occurs on 3.2.3.

**Comment by jira-importer** [comment]
**[Jason Mihalick](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jrmihalick)** commented

[Christian Schulte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=schulte77), so was there any change made?  It looks like commit 020e35816f184c10c3f87f103336fed4516f7af6 made a change, then that change was backed out by 5f048234ff44dbf70fcad9f17834c64866f452e1? We are unable to deploy artifacts to our nexus repository via the Maven release plugin without the workarounds discussed in MNG-5939.

**Comment by jira-importer** [comment]
**[Christian Schulte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=schulte77)** commented

Pre 3.2.3 behaviour got restored in commit [5f048234ff44dbf70fcad9f17834c64866f452e1](https://git-wip-us.apache.org/repos/asf?p=maven.git;a=commitdiff;h=5f048234ff44dbf70fcad9f17834c64866f452e1)

**Comment by jira-importer** [comment]
**[Jason Mihalick](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jrmihalick)** commented

Ok, thank you.  I built maven from the latest 3.4.0-SNAPSHOT and I can confirm that the problem I was having when deploying artifacts to Nexus using the release plugin did not occur when building my project using maven 3.4.0-SNAPSHOT.  Nice job.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Hi,
i have created a example project which shows the behaviour has not being changed for all Maven versions (3.0.5, 3.1.1, 3.2.5, 3.3.9 and current on git master)..
https://github.com/khmarbaise/mshade/tree/master/mshade-195

In the project there are also checked in the log files that i have produced to see what is happening...
So either this issue is not solved or the root cause is something different...

**Comment by jira-importer** [comment]
**[Christian Schulte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=schulte77)** commented

This is not a Maven core issue.  The 'maven-sources-plugin' attaches the source artifact with type 'java-source' and the 'maven-shade-plugin' attaches the source artifact with type 'jar'. That really are two different artifacts. The 'maven-shade-plugin' needs to use the same type as the 'maven-source-plugin' does.

'AbstractSourceJarMojo.java':

```
/**
 * @return The type {@code java-source}
 */
protected String getType()
{
    return "java-source";
}

...

if ( attach )
{
  projectHelper.attachArtifact( project, getType(), getClassifier(), outputFile );
}
```

'ShadeMojo.jar':

```
if ( createSourcesJar )
{
  projectHelper.attachArtifact( project, "jar", shadedClassifierName + "-sources", sourcesJar );
}

...

projectHelper.attachArtifact( project, "jar", "sources", shadedSources );
```

This needs to be changed to read 'java-source' instead of 'jar'. You can easily set a breakpoint at 'MavenProject.addAttachedArtifact' and see what is going on yourself.

**Comment by jira-importer** [comment]
**[Stephen Connolly](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stephenc)** commented

Maven 3.4.0 has been dropped. See [this thread](https://mail-archives.apache.org/mod_mbox/maven-dev/201701.mbox/%3CCA%2BnPnMx-e7kGYy3Hp87v8hLGdhp1q%3DtKLx_6QuZ4kGUqHEBGcw%40mail.gmail.com%3E) for more details.

This issue will need to be re-scheduled for a Maven release in the (hopefully near) future.

**Comment by jira-importer** [comment]
**[Knut Vidar Siem](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=knutsiem)** commented

[Christian Schulte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=schulte77), when you say "this is not a Maven core issue" I assume you're referring to Maven Shade Plugin's use of  "java-source" as source artifact type vs "jar" in Maven Source Plugin. Commit [5f048234ff44dbf70fcad9f17834c64866f452e1](https://git-wip-us.apache.org/repos/asf?p=maven.git;a=commitdiff;h=5f048234ff44dbf70fcad9f17834c64866f452e1) or something to that effect _is_ coming to 3.5.1, right?

```
[MNG-5868] Adding serval times the same artifact via MavenProjectHelper (attachArtifact) does not produce a failure

o Updated to restore the behaviour consensus had been
  reached in MNG-5387.
o Kept 'MavenProject.getAttachedArtifacts' to return an
  unmodifiable list.
```

**Comment by jira-importer** [comment]
**[Christian Schulte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=schulte77)** commented

[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise) will be working on this issue in the future. My initial attempt was to simply throw an exception whenever an existing artifact is attached again. That is what [Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise) had in mind as well. After doing this I discovered quite a few issues in JIRA and just restored that behaviour. I cannot tell what the final solution will be.

**Comment by jira-importer** [comment]
**[Stephen Connolly](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stephenc)** commented

Noting for the record, I do not believe that attaching the same artifact coordinates multiple times should ever produce a failure, as such I view the issue title as incorrect.

This issue seems to be around attaching artifacts using different coordinates but which map to the same effective coordinate.

If you have a custom artifact resolver, the type that an artifact ends up being deployed as may not be the same as the type with which it was attached, hence the simple map that stores the attached artifacts is unaware of the potential duplicates and consequently we end up with two artifacts for the same path. Harmless for the install plugin, and harmless for the deploy plugin when redeployment of artifacts is permitted by the remote repository, but a lot of repositories prevent artifact redeployment and the result is a break for them.

Could probably decompose this into two issues:

1. The shade plugin should be consistent with the javadoc and sources plugins in how it attaches javadoc and sources artifacts

2. **Either** the deploy plugin should dedup the effective list and only deploy the last artifact for any target path **or** core should dedup the artifacts list when attaching artifacts.

At no stage should attaching an artifact trigger a build failure IMHO

**Comment by jira-importer** [comment]
**[Christopher Kohle](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ckohle)** commented

Please note that this also happens with the maven jar plugin when deploying a test jar.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

[Christopher Kohle](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ckohle) If this happens with test-jar I assume you have defined an execution block ?

**Comment by jira-importer** [comment]
**[Christopher Kohle](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ckohle)** commented

Yes,  I'm using an execution block with phase: package goal: test-jar.  Please note that version 3.2.2 of maven works as expected.

**Comment by jira-importer** [comment]
**[Chris Lott](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=chrislott)** commented

I don't mean to rub salt into wounds, but this has been open for almost 5 years now. I see a fix version of 3.7.0-candidate posted here, does that mean there's hope?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

This has been addressed in:

```
$ git log --grep=MNG-5868
commit bc9e7eb687be7cec8916d03a43ebe07b79cc7562
Author: olivier lamy <olamy@apache.org>
Date:   2020-06-01T04:21:13+02:00

    [MNG-5868] No duplicate artifacts in attached artifacts
    if attached artifacts already contains the artifact remove it and add the new one

    Signed-off-by: olivier lamy <olamy@apache.org>

commit 295c205a65b20a9ce53e73ce58f1dd1744b8b542
Author: olivier lamy <olamy@apache.org>
Date:   2020-06-01T04:21:13+02:00

    [MNG-5868] No duplicate artifacts in attached artifacts
    if attached artifacts already contains the artifact remove it and add the new one

    Signed-off-by: olivier lamy <olamy@apache.org>
```

**Comment by jira-importer** [comment]
**[Vincent Massol](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vmassol)** commented

[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy) thanks for the fix. Any problem in backporting it to a 3.6.4 version (since 4.0.0 is not out yet and it may take a while before it is)?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Vincent Massol](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vmassol), there are tentative plans for 3.6.4 or 3.7.0. We will go through the issue list in 4.0.0 and decide which changes we will backport. I am open to bugfixes and non breaking changes unless the list of issues hits a high two-figure number.

**Comment by jira-importer** [comment]
**[Vincent Massol](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vmassol)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) thanks for the reply. This issue could be a good candidate for a 3.6.4 since it's a regression (the worst of bugs ;)).

**Comment by jira-importer** [comment]
**[Vincent Massol](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vmassol)** commented

[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy) Thanks. For us it evens fails our releases since Nexus doesn't accept the duplicate source artifact upload (except when going through staging but that's a pro nexus feature).

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Vincent Massol](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vmassol), can you mean while apply this patch to maven-3.6.x branch locally and work with that in your environment?

**Comment by jira-importer** [comment]
**[Vincent Massol](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vmassol)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) I'm actually not the one who has had the problem. It's some external contributor to some xwiki extension. So it's a bit more complex: I'd need to build maven for them, and ask them to reproduce. It's doable ofc but not immediate.

**Comment by jira-importer** [comment]
**[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy)** commented

I have cherry to maven-3.6.x branch

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) do you have any idea why this log history is such a big mess?? 

 

```java
$ git log --grep=MNG-5868
commit bc9e7eb687be7cec8916d03a43ebe07b79cc7562
Author: olivier lamy <olamy@apache.org>
Date:   2020-06-01T04:21:13+02:00

    [MNG-5868] No duplicate artifacts in attached artifacts
    if attached artifacts already contains the artifact remove it and add the new one

    Signed-off-by: olivier lamy <olamy@apache.org>

commit 295c205a65b20a9ce53e73ce58f1dd1744b8b542
Author: olivier lamy <olamy@apache.org>
Date:   2020-06-01T04:21:13+02:00

    [MNG-5868] No duplicate artifacts in attached artifacts
    if attached artifacts already contains the artifact remove it and add the new one

    Signed-off-by: olivier lamy <olamy@apache.org> 
```

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Why did you change the fix version? 3.6.3 is wrong.

**Comment by jira-importer** [comment]
**[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) oops fixed

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Thanks, LGTM now.

**Comment by jira-importer** [comment]
**[Valentin Delaye](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jonesbusy)** commented

Hi,

It seems this issue also happen with Maven 3.8.1. Any plan to have it available for 3.8.2 ?

Regards,

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [66a43813770cd7dee9eb1ed0372e3f784ba0c541](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=66a43813770cd7dee9eb1ed0372e3f784ba0c541) for `maven-3.8.x` branch.
